kernel/src/fs/utils/inode.rs

```diff
@@ -434,6 +495,45 @@ pub trait Inode: Any + Sync + Send {
     fn extension(&self) -> Option<&Extension> {
         None
     }
+
+    ///  Used to check for read/write/execute permissions on a file.
+    ///
+    ///  We use "fsuid" for this, letting us set arbitrary permissions
+    ///  for filesystem access without changing the "normal" uids which
+    ///  are used for other things..
+    fn check_permission(&self, mut req: RequestPermission) -> Result<()> {
+        if let Some(posix_thread) = current_thread!().as_posix_thread() {
+            req = req.intersection(
+                RequestPermission::MAY_READ
+                    | RequestPermission::MAY_WRITE
+                    | RequestPermission::MAY_EXEC,
+            );
+            let creds = posix_thread.credentials();
+            let mode = self.mode().unwrap();
+
+            if self.metadata().uid == creds.fsuid() {
+                if (req.request_read() && !mode.is_owner_readable())
+                    || (req.request_write() && !mode.is_owner_writable())
+                    || (req.request_exec() && !mode.is_owner_executable())
+                {
+                    return_errno!(Errno::EACCES);
+                }
+            } else if self.metadata().gid == creds.fsgid() {
+                if (req.request_read() && !mode.is_group_readable())
+                    || (req.request_write() && !mode.is_group_writable())
+                    || (req.request_exec() && !mode.is_group_executable())
+                {
+                    return_errno!(Errno::EACCES);
+                }
+            } else if (req.request_read() && !mode.is_other_readable())
+                || (req.request_write() && !mode.is_other_writable())
+                || (req.request_exec() && !mode.is_other_executable())
+            {
+                return_errno!(Errno::EACCES);
+            }
+        };
+        Ok(())
+    }
```

Comment:
Is it correct and aligned with Linux for each type of inode to have its own `check_permission()`? If not, then this function might be better placed in `impl dyn Inode { ... }` to provide a single generic implementation.

Reply:
Yes, in Linux, each file system can have its own implementation of check_permission(), and it also provides a default implementation. Here is an example:

[Default implementation](https://elixir.bootlin.com/linux/v2.6.11/source/fs/namei.c#L167) 。

ext3: [Link to source](https://elixir.bootlin.com/linux/v2.6.11/source/fs/ext3/file.c#L129)

```
struct inode_operations ext3_file_inode_operations = {
    .truncate    = ext3_truncate,
    .setattr     = ext3_setattr,
#ifdef CONFIG_EXT3_FS_XATTR
    .setxattr    = generic_setxattr,
    .getxattr    = generic_getxattr,
    .listxattr   = ext3_listxattr,
    .removexattr = generic_removexattr,
#endif
    .permission  = ext3_permission,
};
```
JFS: [Link to source](https://elixir.bootlin.com/linux/v2.6.11/source/fs/jfs/file.c#L102)


```
struct inode_operations jfs_file_inode_operations = {
    .truncate    = jfs_truncate,
    .setxattr    = jfs_setxattr,
    .getxattr    = jfs_getxattr,
    .listxattr   = jfs_listxattr,
    .removexattr = jfs_removexattr,
#ifdef CONFIG_JFS_POSIX_ACL
    .setattr     = jfs_setattr,
    .permission  = jfs_permission,
#endif
};
```